### PR TITLE
[batch] Inter-Job File Dependencies

### PR DESCRIPTION
**Pull request by danking**
# Batch Inter-Job File Dependencies

The important addition of this PR is a `copy_service_account_name` field on batch jobs that permits the client to authorize with some GCP credentials stored in a k8s secret.

The copy pods and the main pod only share the `/io` folder (k8s forbids mounting a volume as `/`).

This interface is pretty onerous. Pipelines will be updated to use this interface. Pipelines is becoming the easy-to-use interface to batch.

### CODE REVIEW

**Comment by danking** [comment]
There's currently only one matching secret, it's in the test namespace, and it only has permission to write to gs://hail-ci-0-1-batch-volume-test-bucket

File: batch/batch/api.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def __init__(self, timeout=60):
         self.timeout = timeout
 
     def create_job(self, url, spec, attributes, batch_id, callback, parent_ids,
-                   scratch_folder, input_files, output_files):
+                   scratch_folder, input_files, output_files, copy_service_account_name):
```
Comment (cseed) [inline batch/batch/api.py]: So I feel like this isn't quite the right direction, although if you want it as an intermediate step, that's fine.

Obviously the user can't get to specify their service account.  It needs to come from their authenticated credentials.  I haven't looked at the final authn design, but the user account needs to come from claims in the cookie so batch either needs to look at the cookie (or headers created by authn after verifying the cookie).

File: batch/test/test_dag.py [hunk 2]
```diff
@@ -304,3 +304,63 @@ def test_no_parents_allowed_without_batches(client):
...
+def test_output_files_no_service_account_is_error(client):
+    batch = client.create_batch()
+    try:
+        batch.create_job('alpine:3.8',
+                         command=['/bin/sh', '-c', 'echo head > /out'],
+                         output_files=[('/out', 'gs://hail-ci-0-1-batch-volume-test-bucket')])
+    except requests.exceptions.HTTPError as err:
+        assert err.response.status_code == 400
+        assert re.search('.*invalid request: service account may be specified '
+                         'if and only if input_files or output_files is set.*',
```
Comment (jigold) [inline batch/test/test_dag.py]: Do you mean `service account must be specified`

File: batch/test/test_dag.py [hunk 3]
```diff
@@ -304,3 +304,63 @@ def test_no_parents_allowed_without_batches(client):
...
+
+def test_service_account_no_files_is_error(client):
+    batch = client.create_batch()
+    try:
+        batch.create_job('alpine:3.8',
+                         command=['/bin/sh', '-c', 'echo head > /out'],
+                         copy_service_account_name='batch-volume-tester')
+    except requests.exceptions.HTTPError as err:
+        assert err.response.status_code == 400
+        assert re.search('.*invalid request: service account may be specified '
```
Comment (jigold) [inline batch/test/test_dag.py]: Why do we have such a strict rule on this? Why is it bad if I specify the service account but no other files?

File: batch/batch/api.py (hunk 1 above)
Comment (danking) [reply batch/batch/api.py]: Yeah, I'm punting on proper authentication. The job spec is mostly unaudited, the user could specify any volume mounts or service accounts for the main pod they like. I'm just continuing to be embarrassingly permissive.

When we figure out authentication, we'll have the tool to also lock down the job spec.

My current short term thinking is to issue a pair of known random 64 character strings to Konrad and Liam which will serve as their passwords. I think that cribs us until we figure out a proper authentication flow involving oauth

File: batch/test/test_dag.py (hunk 2 above)
Comment (danking) [reply batch/test/test_dag.py]: Hmm. "X may be true if and only if Y is true" sounds right to me. I suppose must also reads fine? I'm not really sure.

File: batch/test/test_dag.py (hunk 3 above)
Comment (danking) [reply batch/test/test_dag.py]: If no files are specified, the copy tasks are not created and the service account is unused, so that seems like a bug.

**Comment by danking** [comment]
Back to you jackie gold.

File: batch/batch/server/server.py [hunk 4]
```diff
@@ -81,15 +81,29 @@ def make_logger():
...
-
-            spec = kube.client.V1PodSpec(containers=[container],
-                                         restart_policy='Never')
-
+    def copy_task(job_id, task_name, files, copy_service_account_name):
+        if files is not None:
+            assert copy_service_account_name is not None
+            authenticate = 'gcloud -q auth activate-service-account --key-file=/gcp-sa-key/key.json'
+            copies = ' & '.join([f'gsutil cp {src} {dst}' for (src, dst) in files])
+            wait = 'wait'
```
Comment (jigold) [inline batch/batch/server/server.py]: What is the `wait` for?

File: batch/batch/server/server.py [hunk 5]
```diff
@@ -81,15 +81,29 @@ def make_logger():
...
+            sh_expression = f'{authenticate} && ({copies} ; {wait})'
+            container = kube.client.V1Container(
+                image='google/cloud-sdk',
```
Comment (jigold) [inline batch/batch/server/server.py]: Do we want to specify the alpine version?

File: batch/batch/server/server.py [hunk 6]
```diff
@@ -81,15 +81,29 @@ def make_logger():
...
-                                                command=['echo', 'hello'])
-
-            spec = kube.client.V1PodSpec(containers=[container],
-                                         restart_policy='Never')
-
+    def copy_task(job_id, task_name, files, copy_service_account_name):
+        if files is not None:
+            assert copy_service_account_name is not None
+            authenticate = 'gcloud -q auth activate-service-account --key-file=/gcp-sa-key/key.json'
+            copies = ' & '.join([f'gsutil cp {src} {dst}' for (src, dst) in files])
```
Comment (jigold) [inline batch/batch/server/server.py]: The other option I looked into was using the `-m` option and giving `cp` a list of files to copy with `-I`. But that's not going to work bc it assumes the same destination directory.

File: batch/batch/server/server.py (hunk 4 above)
Comment (danking) [reply batch/batch/server/server.py]: Each of those `gsutil` commands runs in the background (nb the `' & '`), `wait` with no arguments [waits for all child processes to complete](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/wait.html).

File: batch/batch/server/server.py (hunk 5 above)
Comment (danking) [reply batch/batch/server/server.py]: good point, I also baked in a version number.

File: batch/batch/server/server.py (hunk 4 above)
Comment (jigold) [reply batch/batch/server/server.py]: oh cool! I prob should have known that.

**Pull request merged**